references/depth/stereo/train.py

```diff
@@ -0,0 +1,750 @@
+            )
+        else:
+            raise ValueError(f"Unknown lr decay method {args.lr_decay_method}")
+        schedulers.append(decay_lr_scheduler)
+
+    scheduler = torch.optim.lr_scheduler.SequentialLR(
+        optimizer, schedulers, milestones=milestones
+    )
+    return scheduler
+
+
+def shuffle_dataset(dataset):
+    """Shuffle the dataset"""
+    perm = torch.randperm(len(dataset))
+    return torch.utils.data.Subset(dataset, perm)
+
+
+def resize_dataset_to_n_steps(
+    dataset: torch.utils.data.Dataset, dataset_steps: int, samples_per_step: int, args: argparse.Namespace
+) -> torch.utils.data.Dataset:
+    original_size = len(dataset)
+    if args.steps_is_epochs:
+        samples_per_step = original_size
+    target_size = dataset_steps * samples_per_step
+
+    dataset_copies = []
+    n_expands, remainder = divmod(target_size, original_size)
+    for idx in range(n_expands):
+        dataset_copies.append(dataset)
+
+    if remainder > 0:
+        dataset_copies.append(torch.utils.data.Subset(dataset, list(range(remainder))))
+
+    if args.dataset_shuffle:
+        dataset_copies = [shuffle_dataset(dataset_copy) for dataset_copy in dataset_copies]
+
+    dataset = torch.utils.data.ConcatDataset(dataset_copies)
+    return dataset
+
+
+def get_train_dataset(dataset_root: str, args: argparse.Namespace) -> torch.utils.data.Dataset:
+    datasets = []
+    for dataset_name in args.train_datasets:
+        transform = make_train_transform(args)
+        dataset = make_dataset(dataset_name, dataset_root, transform)
+        datasets.append(dataset)
+
+    if len(datasets) == 0:
+        raise ValueError("No datasets specified for training")
+
+    samples_per_step = args.world_size * args.batch_size
+
+    for idx, (dataset, steps_per_dataset) in enumerate(zip(datasets, args.dataset_steps)):
+        datasets[idx] = resize_dataset_to_n_steps(dataset, steps_per_dataset, samples_per_step, args)
+
+    dataset = torch.utils.data.ConcatDataset(datasets)
+    if args.dataset_order_shuffle:
+        dataset = shuffle_dataset(dataset)
+
+    print(f"Training dataset: {len(dataset)} samples")
+    return dataset
+
+
+@torch.inference_mode()
+def _evaluate(
+    model, args, val_dataset, *, padder_mode, print_freq=10, writter=None, step=None, iterations=None, batch_size=None, header=None
+):
+    """Helper function to compute various metrics (epe, etc.) for a model on a given dataset.
+    We process as many samples as possible with ddp.
+    """
+    model.eval()
+    header = header or "Test:"
+    device = torch.device(args.device)
+    metric_logger = utils.MetricLogger(delimiter="  ")
+    
+    effective_batch_size = batch_size * args.world_size
+    can_do_distributed_test = effective_batch_size < len(val_dataset)
+    
+    if args.distributed and can_do_distributed_test:
+        sampler = torch.utils.data.distributed.DistributedSampler(val_dataset, shuffle=False, drop_last=True)
```

Comment:
question, why we use `drop_last=True`?

Reply:
In order to avoid strange behaviours when the test size is sufficiently large to fit on a step of DDP, but not large enough to be distributed to all worker nodes.

Reply:
Hmm... if the case that you mention happen, isn't drop last gonna drop the whole dataset in that case?
Also, I think you already try to handle this with `can_do_distributed_test` variable, do we need both `drop_last=True` and `can_do_distributed_test` to handle such case?

Reply:
For instance, you have an effective batch_size of 10 and a dataset size of 11. You're going to process the first batch, which will leave you with 1 sample remaining. That sample will be skipped.

`can_do_distributed_test` checks wether or not the dataset size is smaller than the effective batch_size.

Reply:
From the docs: https://pytorch.org/docs/stable/data.html#torch.utils.data.distributed.DistributedSampler
Seems like if we set `drop_last=False` it will add extra indices to make sure the data is evenly divisible across replica.
So I think both have their own issue (not using the whole test data, or having duplicate test data).
In my personal preference though, I prefer to have duplicate test data compared to dropping some of the test data.